python/kvikio/nvcomp.py

```diff
@@ -0,0 +1,224 @@
+# Copyright (c) 2021-2022, NVIDIA CORPORATION. All rights reserved.
+# See file LICENSE for terms.
+
+from enum import Enum
+
+import cupy as cp
+import numpy as np
+
+import kvikio._lib.pynvcomp as _lib
+
+
+def cp_to_nvcomp_dtype(in_type: cp.dtype) -> Enum:
+    """ Convert np/cp dtypes to nvcomp integral dtypes.
+
+    Returns
+    -------
+    int
+        The value of the NVCOMP_TYPE for supported dtype.
+    """
+    cp_type = cp.dtype(in_type)
+    return {
+        cp.dtype("int8"): _lib.pyNvcompType_t.pyNVCOMP_TYPE_CHAR,
+        cp.dtype("uint8"): _lib.pyNvcompType_t.pyNVCOMP_TYPE_UCHAR,
+        cp.dtype("int16"): _lib.pyNvcompType_t.pyNVCOMP_TYPE_SHORT,
+        cp.dtype("uint16"): _lib.pyNvcompType_t.pyNVCOMP_TYPE_USHORT,
+        cp.dtype("int32"): _lib.pyNvcompType_t.pyNVCOMP_TYPE_INT,
+        cp.dtype("uint32"): _lib.pyNvcompType_t.pyNVCOMP_TYPE_UINT,
+        cp.dtype("int64"): _lib.pyNvcompType_t.pyNVCOMP_TYPE_LONGLONG,
+        cp.dtype("uint64"): _lib.pyNvcompType_t.pyNVCOMP_TYPE_ULONGLONG,
+    }[cp_type]
+
+
+class CascadedOptions:
+    """ Options to pass to the Cascaded Compressor.
+
+    This is needed if you want to specify a different number of RLEs or deltas for the
+    Cascaded Compressor.
+    """
+
+    def __init__(self, num_RLEs: int = 1, num_deltas: int = 1, use_bp: bool = True):
+        self.num_RLEs = num_RLEs
+        self.num_deltas = num_deltas
+        self.use_bp = use_bp
+
+
+class CascadedCompressor:
+    def __init__(self, dtype: cp.dtype, config: CascadedOptions = CascadedOptions()):
+        """ Initialize a CascadedCompressor and Decompressor for a
+        specific dtype.
+
+        Parameters
+        ----------
+        dtype: cp.dtype
+            The dtype of the input buffer to be compressed.
+        config: CascadedOptions
+            A CascadedOptions object containing the RLE, deltas, and bp configuration
+            for a CascadedCompressor.
+        """
+        self.dtype = dtype
+        self.config = config
+        self.compressor = _lib._CascadedCompressor(
+            cp_to_nvcomp_dtype(self.dtype).value,
+            config.num_RLEs,
+            config.num_deltas,
+            config.use_bp,
+        )
+        self.decompressor = _lib._CascadedDecompressor()
+        self.s = cp.cuda.Stream()
```

Comment:
I think immutable is a good idea here, the fact that they are properties of the class is really just making them available as queryable parameters. They've been set in the underlying C++ class and are no longer accessible by any API, so I keep a record here.

Reply:
Note for future work: Depending on whether users might actually need to access these properties, I would recommend either renaming them with underscores (e.g. `self._decompressor`) or turning them into properties instead of raw attributes.